examples/routing_service/routing_service_file_adapter/cpp/FileStreamReader.cxx

```diff
@@ -0,0 +1,129 @@
+/*****************************************************************************/
+/*         (c) Copyright, Real-Time Innovations, All rights reserved.        */
+/*                                                                           */
+/*         Permission to modify and use for internal purposes granted.       */
+/* This software is provided "as is", without warranty, express or implied.  */
+/*                                                                           */
+/*****************************************************************************/
+
+#include <sstream>
+#include <thread>
+
+#include "FileStreamReader.hpp"
+
+using namespace rti::community::examples;
+using namespace dds::core::xtypes;
+
+const std::string FileStreamReader::INPUT_FILE_PROPERTY_NAME = "example.adapter.input_file";
+const std::string FileStreamReader::SAMPLE_PERIOD_PROPERTY_NAME = "example.adapter.sample_period_sec";
+
+void FileStreamReader::ProcessThread()
+{
+    stop_thread_ = false;
+
+    while (!stop_thread_) {
+        if (inputfile_.is_open()) {
+            std::getline(inputfile_, buffer_);
+
+            // Here we notify routing service, that there is data available
+            // on the StreamReader, triggering a call to take().
+            if (!inputfile_.eof()) {
+                reader_listener_->on_data_available(this);
+            }
+        }
+
+        std::this_thread::sleep_for(std::chrono::seconds(sampling_period_));
+    }
+}
+
+FileStreamReader::FileStreamReader(
+        const StreamInfo &info,
+        const PropertySet &property,
+        StreamReaderListener *listener)
+        : filereader_thread_(),
+        sampling_period_(1)
+{
+    reader_listener_ = listener;
+    adapter_type_ = static_cast<DynamicType *>(
+            info.type_info().type_representation());
+
+    // Parse the properties provided in the xml configuration file
+    for (const auto &it : property) {
+        if (it.first == INPUT_FILE_PROPERTY_NAME) {
+            inputfile_.open(it.second);
+        } else if (it.first == SAMPLE_PERIOD_PROPERTY_NAME) {
+            sampling_period_ = std::stoi(it.second);
+        }
+    }
+
+    if (!inputfile_.is_open()) {
+        throw std::logic_error("Input file not provided or unable to open");
+    }
+
+    filereader_thread_ = std::thread(&FileStreamReader::ProcessThread, this);
+}
+
+void FileStreamReader::read(
+        std::vector<dds::core::xtypes::DynamicData *> &samples,
+        std::vector<dds::sub::SampleInfo *> &infos)
+{
+    take(samples, infos);
```

Comment:
implementing read is not required. Default implementation will throw an exception. So if we don't intentd to implement actually the read, we should leave with default implementation. For basic forwarding (when default processor is working), only take(data, info) is required.

Reply:
Done.